### ISSUE TITLE
Question about the loss computation in the Trainer class when we need gradient accumulation.

### ISSUE THREAD

**Issue by AlfredQin** [Deci-AI/super-gradients#1713]
### 💡 Your Question

Hi, I have a question about the gradient accumulation in the SG Trainer class. I see others often divide the loss by the number of accumulation steps before backward. It seems that there is no division for the loss in the SG trainer. So  I want to know if it is a bug or maybe I have misunderstanding.

**Comment by BloodAxe** [comment]
The "gradient accumulation" term suggests that you are accumulating (adding up) gradients, and not averaging them. It is new thing to me to hear someone is dividing accumulated gradients by accumulation step.  
Our implementation aligns with this discussion on pytorch forum in that regard:
https://discuss.pytorch.org/t/why-do-we-need-to-set-the-gradients-manually-to-zero-in-pytorch/4903?u=alband

In SG we accumulate the gradients and not dividing grads or loss by the accumulation size.
Some libraries (Like HF Accelerate) do [this](https://huggingface.co/docs/accelerate/usage_guides/gradient_accumulation) however. One remark I was able to find regarding this division is from https://kozodoi.me/blog/20210219/gradient-accumulation:

> Divide the running loss by `acum_iter`. This normalizes the loss to reduce the contribution of each mini-batch we are actually processing. **Depending on the way you compute the loss, you might not need this step: if you average loss within each batch, the division is already correct and there is no need for extra normalization**.

My understanding is this: if we do not divide the loss by accumulation factor we essentially multiply the learning rate by that number. Assuming we use accumulation factor of 4 and our optimized can produce perfect gradients towards global minimum:

1) If we are not using gradient accumulation and just run 4 optimization steps one after another, we would would get `g1`, `g2`, `g3`, `g4` gradients.

2) If we are using gradient accumulation without division, we should get (g1 + g2 + g3 + g4) final gradient which should move the weights of the model to the same state as in previous step (Under considerations that we have perfect optimizer and toy problem with global minima)

3) Now let's add the normalization. Not the final gradient would be `(g1 + g2 + g3 + g4) / acc_step` so trainer would have to do more steps (Exactly `acc_step` more) to get to the same state.

So in Option 3 we have to do more steps. In some cases this may be beneficial because the step directions will be less "noisy". Alternative approach is to lower the learning rate by same factor.

How this answers your question.

**Comment by AntonioConsiglio** [comment]
The difference between small batch and large batch lies in the fact that a small batch does not represent our dataset statistically and therefore we may have more fluctuations between one batch step and the next, is it right? Based on this assumption, accumulating the gradient by going to normalize the loss by dividing it by the number of gradients you want to accumulate leads to having a gradient that is representative of a larger batch, which leads to us not having the fluctuations that we might have in the case where we are using a small batch. From your explanation @BloodAxe it does not show this, and especially in point 2, the gradient g2, g3, g4 are different than in case 3 because the network weights are changed at each step.
